sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs

```diff
@@ -1059,6 +1059,15 @@ public virtual async Task FlushAsync(CancellationToken cancellationToken = defau
 
                 await StopPublishingAsync(cancelActiveSendOperations: false, cancellationToken).ConfigureAwait(false);
                 await FlushInternalAsync(cancellationToken).ConfigureAwait(false);
+
+                // There's an unlikely race condition where it is possible that an event batch was being enqueued before publishing
+                // stopped and was not written to the partition buffer until after the flush for that partition completed.  To guard
+                // against this, restart publishing if any events are pending.
+
+                if (Volatile.Read(ref _totalBufferedEventCount) > 0)
+                {
+                    await StartPublishingAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I know this is a rare scenario but it feels like it changes the behavior of the method a bit since before background processes were turned off through stop publishing async? Although, flush doesn't commit to stopping the publishing so it's probably ok to change since it's an implementation detail?

Also, does this edge case impact `CloseAsync` as well?

Reply:
Yeah, it's a behavioral change, for sure.   But, it fixes a bug with another race corner case:

- Enqueue gets called and gets past the `IsPublishing` check (L936) then gets paused before writing the event.
- Flush gets called, grabs the semaphore and stops publishing, clears all the buffers
- Enqueue resumes and writes the event to the buffer
- Nobody else calls enqueue 

That leaves an orphaned event in the buffer with publishing not running.   In that scenario, we violate our contract because we're not honoring the maximum wait time.   The event doesn't get published until someone calls Enqueue, Flush, or Close.

Reply:
> Also, does this edge case impact CloseAsync as well?

It would, but we structured that one correctly from the start.   The semaphore acquisition is outside the try/finally block that releases it, so cancellation won't cause a release.